File: durak/src/rules.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

use rand::seq::SliceRandom;
use rand::{thread_rng, seq::IteratorRandom};

#[derive(Serialize, Deserialize, Clone, Copy, Debug)]
pub enum Verb {
    Play = 1,
    Cover,
    Reverse,
    Pass,
    PickUp,
    Defer,
}

pub type Card = usize;
pub const UNK_CARD : Card = 36;

pub fn new_card(card: usize) -> Card {
    card
}

pub fn card_from_suit_rank(suit: usize, rank: usize) -> Card {
    suit * 9 + rank
}

pub fn card_suit(card: Card) -> usize {
    card / 9
}

pub fn card_rank(card: Card) -> usize {
    card % 9
}

pub fn card_suit_string(card: Card) -> String {
    match card_suit(card) {
        0 => "spades".to_string(),
        1 => "hearts".to_string(),
        2 => "diamonds".to_string(),
        3 => "clubs".to_string(),
        _ => "unknown".to_string(),
    }
}

pub fn card_rank_string(card: Card) -> String {
    match card_rank(card) {
        0 => "6".to_string(),
        1 => "7".to_string(),
        2 => "8".to_string(),
        3 => "9".to_string(),
        4 => "10".to_string(),
        5 => "J".to_string(),
        6 => "Q".to_string(),
        7 => "K".to_string(),
        8 => "A".to_string(),
        _ => "unknown".to_string(),
    }
}

pub fn card_to_string(card: Card) -> String {
    format!("{} of {}", card_rank_string(card), card_suit_string(card))
}

pub fn beats(card1: Card, card2: Card, trump: Card) -> bool {
    if card_suit(card1) == card_suit(trump) && card_suit(card1) != card_suit(card2) {
        return true
    }
    card_rank(card1) > card_rank(card2) && card_suit(card1) == card_suit(card2)
}
    
pub fn generate_deck() -> Vec<Card> {
    let mut res = Vec::new();
    for suit in 0..4 {
        for rank in 0..9 {
            res.push(card_from_suit_rank(suit, rank));
        }
    }
    res.shuffle(&mut thread_rng());
    res
}

pub fn remove_card(cards: &mut Vec<Card>, card: Card) -> bool {
    let idx = cards.iter().position(|x| *x == card);
    match idx {
        Some(i) => {
            cards.remove(i);
            true
        },
        None => false,
    }
}

#[derive(Serialize, Deserialize, Clone, Copy, Debug)]
pub struct Action {
    pub player: usize,
    pub verb: Verb,
    pub card: Card,
    pub covering: Card,
}

impl Action {
    pub fn new(player: usize, verb: Verb, card: Card, covering: Card) -> Action {
        Action {
            player,
            verb,
            card,
            covering,
        }
    }
}

#[derive(Serialize, Deserialize, Debug)]
pub struct GameState {
    pub defender: usize,
    pub picking_up: bool,
    pub trump: Card,
    pub plays: Vec<Card>,
    pub covers: Vec<Card>,
    pub hands: Vec<Vec<Card>>,
    pub defering: bool,
    #[serde(skip)]
    pub start: Option<time::Instant>,
}

pub fn num_not_unk(v: &Vec<Card>) -> usize {
    v.iter().filter(|&x| *x != UNK_CARD).count()
}

impl GameState {
    pub fn new(trump: Card, hands: Vec<Vec<Card>>) -> GameState {
        GameState {
            defender: 1,
            picking_up: false,
            trump,
            plays: Vec::new(),
            covers: Vec::new(),
            hands,
            defering: false,
            start: None,
        } 
    }
    pub fn attacker_actions(&self, pidx: usize) -> Vec<Action> {
        let mut res = Vec::new();
        if self.plays.len() == 0 {
           for &card in &self.hands[pidx] {
               res.push(Action::new(pidx, Verb::Play, card, UNK_CARD))
           } 
        } else {
            for &card in &self.hands[pidx] {
                for &board_card in self.plays.iter().chain(self.covers.iter()) {
                    if board_card == UNK_CARD {
                        continue;
                    }
                    // Card equality with unknown card for AI search
                    if card_rank(card) == card_rank(board_card) || card == UNK_CARD {
                        res.push(Action::new(pidx, Verb::Play, card, UNK_CARD));
                        break;
                    }
                }
            }
        }
        if self.picking_up || (num_not_unk(&self.covers) == self.plays.len() && self.plays.len() > 0) {
            res.push(Action::new(pidx, Verb::Pass, UNK_CARD, UNK_CARD))
        }
        // For AI to not throw trumps away
        if !self.picking_up && self.plays.len() > num_not_unk(&self.covers) {
            res.push(Action::new(pidx, Verb::Defer, UNK_CARD, UNK_CARD)) 
        }
        res
    }
    pub fn reverse_rank(&self) -> Option<usize> {
        if self.plays.len() == 0 && self.covers.len() == 0 {
            return None
        }
        let mut card = self.plays[0];
        for &c in self.plays.iter().chain(self.covers.iter()) {
            if c != UNK_CARD && c != card {
                return None
            }
        }
        return Some(card_rank(card))
    }
    pub fn defender_actions(&self, pidx: usize) -> Vec<Action> {
        let mut res = Vec::new();
        if self.picking_up {
            return res
        }
        let rev_rank = self.reverse_rank();
        match rev_rank {
            Some(rank) => {
                for &card in &self.hands[pidx] {
                    if card_rank(card) == rank {
                        res.push(Action::new(pidx, Verb::Reverse, card, UNK_CARD))
                    }
                }
            },
            None => (),
        }
        for i in 0..self.plays.len() {
            for &card in self.hands[pidx].iter() {
                if self.covers[i] == UNK_CARD && beats(card, self.plays[i], self.trump) {
                    res.push(Action::new(pidx, Verb::Cover, card, self.plays[i]))
                }
            }
        }
        if self.plays.len() > 0 && self.covers.iter().filter(|&x| *x != UNK_CARD).count() < self.plays.len() {
            res.push(Action::new(pidx, Verb::PickUp, UNK_CARD, UNK_CARD))
        }
        res
    }
    pub fn player_actions(&self, pidx: usize) -> Vec<Action> {
        if pidx == self.defender {
            self.defender_actions(pidx)
        } else {
            self.attacker_actions(pidx)
        }
    }
    pub fn random_action(&self) -> Action {
        self.player_actions(0).iter().chain(self.player_actions(1).iter()).choose(&mut thread_rng()).unwrap().clone()
    }
    pub fn take_action(&mut self, action: &Action) {
        match action.verb {
            Verb::Play => {
                self.plays.push(action.card);
                self.covers.push(UNK_CARD);
                remove_card(&mut self.hands[action.player], action.card);
            }, 
            Verb::Cover => {
                for i in 0..self.plays.len() {
                    if action.covering == self.plays[i] {
                        self.covers[i] = action.card;
                    }
                }
                remove_card(&mut self.hands[action.player], action.card);
            },
            Verb::Reverse => {
                self.plays.push(action.card);
                self.covers.push(UNK_CARD);
                remove_card(&mut self.hands[action.player], action.card);
                self.defender = 1-self.defender;
                self.defering = false;
            },
            Verb::PickUp => {
                self.picking_up = true;
            },
            Verb::Pass => {
                if self.picking_up {
                    for &card in self.plays.iter().chain(self.covers.iter()) {
                        if card != UNK_CARD {
                            self.hands[self.defender].push(card);
                        } 
                    } 
                }
                self.plays.clear();
                self.covers.clear();
                self.picking_up = false;
                self.defering = false;
            },
            Verb::Defer => {
                self.defering = true;
            },
            _ => (),
        }
    }
    pub fn clone(&self) -> GameState {
        GameState {
            defender: self.defender,
            picking_up: self.picking_up,
            trump: self.trump,
            plays: self.plays.clone(),
            covers: self.covers.clone(),
            hands: vec![self.hands[0].clone(), self.hands[1].clone()],
            defering: self.defering,
            start: self.start,
        }
    }
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Memory {
    pub hands: Vec<Vec<Card>>,
    pub sizes: Vec<usize>,
    pub discard: Vec<Card>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Game {
    pub key: usize,
    pub state: GameState,
    pub deck: Vec<Card>,
    pub memory: Memory,
    pub recording: Vec<String>,
    pub versus: String,
    pub winner: Option<usize>,
    pub joined: bool,
}

impl Game {
    pub fn new(key: usize, versus: String) -> Game {
        let deck = generate_deck();
        Game {
            key,
            state: GameState::new(deck[0], vec![deck[30..36].to_vec(), deck[24..30].to_vec()]),
            deck: deck[..24].to_vec(),
            memory: Memory {
                hands: vec![Vec::new(), Vec::new()],
                sizes: vec![6, 6],
                discard: Vec::new(),
            },
            recording: Vec::new(),
            versus,
            winner: None,
            joined: false,
        }
    }
    pub fn take_action(&mut self, action: &Action) {
        self.recording.push(serde_json::to_string(&action).unwrap());
        match action.verb {
            Verb::Play | Verb::Cover => {
                self.state.take_action(&action);
                remove_card(&mut self.memory.hands[action.player], action.card);
                self.memory.sizes[action.player] -= 1;
            },
            Verb::Reverse => {
                self.state.take_action(&action);
                remove_card(&mut self.memory.hands[action.player], action.card);
                self.memory.sizes[action.player] -= 1;
            },
            Verb::Pass => {
                if self.state.picking_up {
                    for &card in self.state.plays.iter().chain(self.state.covers.iter()) {
                        if card != UNK_CARD {
                            self.memory.hands[self.state.defender].push(card);
                            self.memory.sizes[self.state.defender] += 1;
                        }
                    }
                } else {
                    for &card in self.state.plays.iter().chain(self.state.covers.iter()) {
                        assert!(card != UNK_CARD);
                        self.memory.discard.push(card);
                    }
                }
                self.state.take_action(&action);
                self.deal(1-self.state.defender);
                self.deal(self.state.defender);
            },
            Verb::PickUp | Verb::Defer => {
                self.state.take_action(&action);
            },
        }
    }
    pub fn deal(&mut self, player: usize) {
        while self.deck.len() > 0 && self.state.hands[player].len() < 6 {
            self.state.hands[player].push(self.deck.pop().unwrap());
        }
        self.memory.sizes[player] = self.state.hands[player].len();
    }
    pub fn is_over(&self) -> bool {
        if self.deck.len() == 0 && self.state.hands.iter().any(|x| x.len() == 0) {
            return true
        }
        false
    }
}

File: v5/rules.go
package main

import (
    "encoding/json"
    "fmt"
    //"log"
    "math/rand"
    "sync"
    "time"
    
    "github.com/gorilla/websocket"
)

type Verb int
type Card int

var UNK_CARD = Card(-1)

const (
    PlayVerb Verb = iota
    CoverVerb 
    ReverseVerb
    PassVerb
    PickupVerb
    DeferVerb
)

var suits = []string{"Clubs", "Spades", "Hearts", "Diamonds"}
var ranks = []string{"6", "7", "8", "9", "10", "Jack", "Queen", "King", "Ace"}
var verbs = []string{"Play", "Cover", "Reverse", "Pass", "Pickup", "Defer"}

func CardFromRankSuit(rank int, suit int) Card {
    return Card(suit*9 + rank)
}

func (card Card) Rank() int {
    return int(card)%9
}

func (card Card) Suit() int {
    return int(card)/9
}

func (card Card) RankStr() string {
    return ranks[card.Rank()]
}

func (card Card) SuitStr() string {
    return suits[card.Suit()]
}

func (card Card) ToStr() string {
    return fmt.Sprintf("%s of %s", card.RankStr(), card.SuitStr()) 
}

func (card Card) Beats(other Card, trump Card) bool {
    if card.Suit() == trump.Suit() && card.Suit() != other.Suit() {
        return true
    }
    return card.Rank() > other.Rank() && card.Suit() == other.Suit()
}

func GenerateDeck() []Card {
    res := make([]Card, 0)
    for suit := 0; suit < 4; suit++ {
        for rank := 0; rank < 9; rank++ {
            res = append(res, CardFromRankSuit(rank, suit))
        }
    }
    rand.Shuffle(len(res), func(i, j int) {
        res[i], res[j] = res[j], res[i]
    })
    return res
}

func RemoveCard(cards *[]Card, c Card) bool {
    for i,card := range *cards {
        if card == c {
            (*cards)[i] = (*cards)[len(*cards)-1]
            *cards = (*cards)[:len(*cards)-1]
            return true
        }
    }
    return false
}

type Action struct {
    Player int
    Verb Verb
    Card Card
    Covering Card
}

func (a Action) IsNull() bool {
    return a.Card == 0 && a.Covering == 0
}

func (a Action) ToStr() string {
    mp := map[string]any {
        "Player": a.Player,
        "Verb": verbs[a.Verb],
        "Card": a.Card,
        "Covering": a.Covering,
    }
    jsn, _ := json.Marshal(mp)
    return string(jsn)
}

type GameState struct {
    Attacker int
    Defender int
    PickingUp bool
    Deferring []bool
    Passed []bool
    Won []bool
    Trump Card
    Plays []Card
    Covers []Card
    Hands [][]Card
    Dir int
    start time.Time
    gamePtr *Game
}

func NumNotUnk(cards []Card) int {
    res := 0
    for _,c := range cards {
        if c != UNK_CARD {
            res++
        }
    }
    return res
}

func InitGameState(trump Card, hands [][]Card) *GameState {
    return &GameState{
        Attacker: 0,
        Defender: 1, 
        PickingUp: false, 
        Deferring: make([]bool, len(hands)),
        Passed: make([]bool, len(hands)),
        Won: make([]bool, len(hands)),
        Trump: trump,
        Plays: make([]Card, 0),
        Covers: make([]Card, 0),
        Hands: hands,
        Dir: 1,
        start: time.Now(),
        gamePtr: nil,
    }
}
    
func (state *GameState) AttackerActions(player int) []Action {
    res := make([]Action, 0)
    // Player has already passed
    if state.Passed[player] {
        return res
    }
    // Player has already won
    if state.Won[player] {
        return []Action{Action{player, PassVerb, UNK_CARD, UNK_CARD}}
    }
    if len(state.Plays) == 0 {
        // Only initial attacker may play first
        if state.Attacker != player {
            return res
        }
        for _,card := range state.Hands[player] {
            res = append(res, Action{player, PlayVerb, card, UNK_CARD})
        }
        return res
    }
    // Don't allow playing more cards than defender can defend
    unmetCards := len(state.Plays) - NumNotUnk(state.Covers)
    if len(state.Hands[state.Defender]) - unmetCards > 0 {
        for _,card := range state.Hands[player] {
            // Allow play unknown card in search
            if card == UNK_CARD {
                res = append(res, Action{player, PlayVerb, card, UNK_CARD})
                continue
            }
            // Regular cards
            for i := 0; i < len(state.Plays); i++ {
                if card.Rank() == state.Plays[i].Rank() || (card != UNK_CARD && card.Rank() == state.Covers[i].Rank()) {
                    res = append(res, Action{player, PlayVerb, card, UNK_CARD})
                    break
                }
            }
        }
    }
    if state.PickingUp || (NumNotUnk(state.Covers) == len(state.Plays) && len(state.Plays) > 0) {
        res = append(res, Action{player, PassVerb, UNK_CARD, UNK_CARD})
    }
    // For AI to not throw trumps away
    if !state.PickingUp && len(state.Plays) > NumNotUnk(state.Covers) {
        res = append(res, Action{player, DeferVerb, UNK_CARD, UNK_CARD})    
    }
    return res
}

func (state *GameState) ReverseRank() int {
    if len(state.Plays) == 0 {
        return -1
    }
    rank := state.Plays[0].Rank()
    for i := 0; i < len(state.Plays); i++ {
        if rank != state.Plays[i].Rank() {
            return -1
        }
        if state.Covers[i] != UNK_CARD && rank != state.Covers[i].Rank() {
            return -1
        }
    }
    return rank
}

func (state *GameState) DefenderActions(player int) []Action {
    res := make([]Action, 0)
    if state.PickingUp {
        return res
    }
    revRank := state.ReverseRank()
    // Only allow reverse when defender can potentially meet it
    if revRank != -1 && NumNotUnk(state.Covers) == 0 && len(state.Plays)+1 <= len(state.Hands[state.Attacker]) {
        for _,card := range state.Hands[player] {
            if card.Rank() == revRank {
                res = append(res, Action{player, ReverseVerb, card, UNK_CARD})
            }
        }
    }
    for i := 0; i < len(state.Plays); i++ {
        for _,card := range state.Hands[player] {
            // For AI search allow cover with unknown card
            if card == UNK_CARD || (state.Covers[i] == UNK_CARD && card.Beats(state.Plays[i], state.Trump)) {
                res = append(res, Action{player, CoverVerb, card, state.Plays[i]})
            }
        }
    }
    if len(state.Plays) > 0 && NumNotUnk(state.Covers) < len(state.Plays) {
        res = append(res, Action{player, PickupVerb, UNK_CARD, UNK_CARD})
    }
    return res
}

func IndexOf[T comparable](slice []T, val T) int {
    for i, v := range slice {
        if v == val {
            return i
        }
    }
    return -1
}

func (state *GameState) PlayerActions(player int) []Action {
    if player == state.Defender {
        return state.DefenderActions(player)
    } else {
        return state.AttackerActions(player)
    }
}

func (state *GameState) RandomAction() Action {
    acts := append(make([]Action, 0), state.PlayerActions(0)...)
    acts = append(acts, state.PlayerActions(1)...)
    return acts[rand.Intn(len(acts))]
}

func (state *GameState) NextRole(player int) int {
    for i := 0; i < len(state.Hands); i++ {
        p := player+((i+1)*state.Dir)
        if p < 0 {
            p += len(state.Hands)
        }
        if p >= len(state.Hands) {
            p -= len(state.Hands)
        }
        if !state.Won[p] {
            return p
        }
    }
    panic("NextRole failed")
}

func (state *GameState) AllPassed() bool {
    for i := 0; i < len(state.Hands); i++ {
        if !state.Passed[i] && state.Defender != i && !state.Won[i] {
            return false
        }
    }
    return true
}

func (state *GameState) CheckGameOver() bool {
    if len(state.gamePtr.Deck) > 0 {
        return false
    }
    n := 0
    for i := 0; i < len(state.Hands); i++ {
        if len(state.Hands[i]) == 0 {
            state.Won[i] = true
            n++
        }
    }
    return n == len(state.Hands)-1
}

func (state *GameState) TakeAction(action Action) {
    switch action.Verb {
        case PlayVerb: {
            state.Plays = append(state.Plays, action.Card)
            state.Covers = append(state.Covers, UNK_CARD)
            RemoveCard(&state.Hands[action.Player], action.Card)
            // Sets to false
            state.Deferring = make([]bool, len(state.Hands))
            state.CheckGameOver()
        }
        case CoverVerb: {
            for i := 0; i < len(state.Plays); i++ {
                if action.Covering == state.Plays[i] {
                    state.Covers[i] = action.Card
                }
            }
            RemoveCard(&state.Hands[action.Player], action.Card)
            state.Deferring = make([]bool, len(state.Hands))
            state.CheckGameOver()
        }
        case ReverseVerb: {
            state.Plays = append(state.Plays, action.Card)
            state.Covers = append(state.Covers, UNK_CARD)
            RemoveCard(&state.Hands[action.Player], action.Card)
            state.Attacker, state.Defender = state.Defender, state.Attacker
            state.Deferring = make([]bool, len(state.Hands))
            state.Dir *= -1
            state.CheckGameOver()
        }
        case PickupVerb: {
            state.PickingUp = true
            state.Deferring = make([]bool, len(state.Hands))
        }
        case PassVerb: {
            // Handled in Game.TakeAction 
            // Need to be in both places for Medium search
            state.Passed[action.Player] = true
            if !state.AllPassed() {
                break
            }
            if state.PickingUp {
                for i := 0; i < len(state.Plays); i++ {
                    state.Hands[state.Defender] = append(state.Hands[state.Defender], state.Plays[i])
                    if state.Covers[i] != UNK_CARD {
                        state.Hands[state.Defender] = append(state.Hands[state.Defender], state.Covers[i])
                    }
                }
                state.Attacker = state.NextRole(state.Defender) 
                state.Defender = state.NextRole(state.Attacker)
            } else {
                state.Attacker = state.NextRole(state.Attacker) 
                state.Defender = state.NextRole(state.Attacker)
            }
            state.Plays = make([]Card, 0)
            state.Covers = make([]Card, 0)
            state.PickingUp = false
            state.Deferring = make([]bool, len(state.Hands))
            state.Passed = make([]bool, len(state.Hands))
        }
        case DeferVerb: {
            state.Deferring[action.Player] = true
        }
    }
}

func (state *GameState) Clone() *GameState {
    hands := make([][]Card, len(state.Hands))
    for i := 0; i < len(hands); i++ {
        hands[i] = append(make([]Card, 0), state.Hands[i]...)
    }
    return &GameState {
        Defender: state.Defender,
        Attacker: state.Attacker,
        PickingUp: state.PickingUp,
        Deferring: append(make([]bool, 0), state.Deferring...),
        Passed: append(make([]bool, 0), state.Passed...),
        Won: append(make([]bool, 0), state.Won...),
        Trump: state.Trump,
        Plays: append(make([]Card, 0), state.Plays...),
        Covers: append(make([]Card, 0), state.Covers...),
        Hands: hands,
        Dir: state.Dir,
        gamePtr: state.gamePtr,
        start: state.start,
    }
}

func (state *GameState) ToStr() string {
    jsn, _ := json.Marshal(state.Clone())
    return string(jsn)
}

type Memory struct {
    Hands [][]Card
    Sizes []int
    Discard []Card
}

type Game struct {
    Key int
    State *GameState
    Deck []Card
    Memory *Memory
    Recording *Recording
    Players []string
    Names []string
    joined []bool
    mutex sync.Mutex
    conns []*websocket.Conn
}

type Recording struct {
    Players []string
    Deck []Card
    Hands [][]Card
    Actions []Action
    Winners []int
}

func InitGame(key int, players []string) *Game {
    deck := GenerateDeck()
    numPlayers := len(players)
    handsState := make([][]Card, numPlayers)
    handsRec := make([][]Card, numPlayers)
    handsMemory := make([][]Card, numPlayers)
    for i := 0; i < numPlayers; i++ {
        handsState[i] = append(make([]Card, 0), deck[i*6:(i+1)*6]...)
        handsRec[i] = append(make([]Card, 0), deck[i*6:(i+1)*6]...)
        handsMemory[i] = make([]Card, 0)
    }
    deck = append(make([]Card, 0), deck[numPlayers*6:]...)
    recording := &Recording{
        Players: players,
        Deck: append(make([]Card, 0), deck...),
        Hands: handsRec,
        Actions: make([]Action, 0),
        Winners: make([]int, 0),
    }
    game := &Game{
        Key: key,
        State: InitGameState(deck[0], handsState),
        Deck: deck,
        Memory: &Memory {
            Hands: handsMemory,
            Sizes: make([]int, numPlayers),
            Discard: make([]Card, 0),
        },
        Recording: recording,
        Players: players,
        Names: make([]string, numPlayers),
        joined: make([]bool, numPlayers),
        conns: make([]*websocket.Conn, numPlayers),
    }
    game.State.gamePtr = game
    return game
}

func (game *Game) CheckGameOver() bool {
    if len(game.Deck) > 0 {
        return false
    }
    if len(game.Recording.Winners) == len(game.Players)-1 {
        return true
    }
    for i := 0; i < len(game.State.Hands); i++ {
        if len(game.State.Hands[i]) == 0 {
            if IndexOf(game.Recording.Winners, i) == -1 {
                game.Recording.Winners = append(game.Recording.Winners, i)
            }
        }
    }
    return len(game.Recording.Winners) == len(game.Players)-1 
}

// TODO Deal in correct order
func (game *Game) Deal(player int) {
   for len(game.Deck) > 0 && len(game.State.Hands[player]) < 6 {
       game.State.Hands[player] = append(game.State.Hands[player], game.Deck[len(game.Deck)-1])
       game.Deck = game.Deck[:len(game.Deck)-1]
   } 
   game.Memory.Sizes[player] = len(game.State.Hands[player])
   // If some player has zero cards and we don't deal them any
   // Adjust winners, attackers, defenders
   game.CheckGameOver()
   if IndexOf(game.Recording.Winners, game.State.Attacker) != -1 {
       game.State.Attacker = game.State.NextRole(game.State.Attacker)
       game.State.Defender = game.State.NextRole(game.State.Attacker)
   }
   if IndexOf(game.Recording.Winners, game.State.Defender) != -1 {
       game.State.Defender = game.State.NextRole(game.State.Defender)
       game.State.Attacker = game.State.NextRole(game.State.Defender)
   }
}

func (game *Game) TakeAction(action Action) bool {
    // Check that game isn't over
    if game.CheckGameOver() {
        return false
    }
    // Check that action is still legal
    acts := game.State.PlayerActions(action.Player)
    if IndexOf(acts, action) == -1 {
        return false
    }
    // No strings of Defer verbs in recordings
    if action.Verb == DeferVerb {
        if len(game.Recording.Actions) > 0 && game.Recording.Actions[len(game.Recording.Actions)-1].Verb != DeferVerb {
            game.Recording.Actions = append(game.Recording.Actions, action)
        }
    } else {
        game.Recording.Actions = append(game.Recording.Actions, action)
    }
    switch action.Verb {
        case PlayVerb, CoverVerb, ReverseVerb: {
            game.State.TakeAction(action);
            RemoveCard(&game.Memory.Hands[action.Player], action.Card)
            game.Memory.Sizes[action.Player] -= 1
        }
        case PassVerb: {
            game.State.Passed[action.Player] = true
            if !game.State.AllPassed() {
                break
            }
            if game.State.PickingUp {
                for i := 0; i < len(game.State.Plays); i++ {
                    game.Memory.Hands[game.State.Defender] = append(game.Memory.Hands[game.State.Defender], game.State.Plays[i])
                    game.Memory.Sizes[game.State.Defender] += 1
                    if game.State.Covers[i] != UNK_CARD {
                        game.Memory.Hands[game.State.Defender] = append(game.Memory.Hands[game.State.Defender], game.State.Covers[i])
                        game.Memory.Sizes[game.State.Defender] += 1
                    }
                }
            } else {
                for i := 0; i < len(game.State.Plays); i++ {
                    game.Memory.Discard = append(game.Memory.Discard, game.State.Plays[i])
                    game.Memory.Discard = append(game.Memory.Discard, game.State.Covers[i])
                }
            }
            game.State.TakeAction(action);
            for i := 0; i < len(game.State.Hands); i++ {
                game.Deal(i)
            }
        }
        case PickupVerb, DeferVerb: {
            game.State.TakeAction(action);
        }
    }
    return true
}
